### ISSUE TITLE
Add `descendant_entered_tree` signal to `Node`

### ISSUE THREAD

**Issue by Arecher** [godotengine/godot-proposals#10711]
### Describe the project you are working on

Working on a custom `Container` with a shader attached, that is meant to affect all it's descendants.

I am currently working in Godot 3.5.stable (soon 3.6.stable), so this proposal and the functions mentioned will be written in that context. To my best knowledge this proposal applies to 4.x as well.

### Describe the problem or limitation you are having in your project

Since I'm using a `ShaderMaterial` in the `material` property of the `CanvasItem`, the only way to get it to apply to all the Container's descendants, is to propagate `use_parent_material` to true in all descendants of the Container.

![image](https://github.com/user-attachments/assets/639a0d1e-4955-4955-b7c4-2646d6dcfcd1)
_All the nodes below `CustomContainer need to use the CustomContainer's ShaderMaterial_

This is no problem, built-in functions such as `propagate_call()` exist, and we can call that on `_ready()`. However, when in during runtime nodes get added to the tree, as descendants of this node, I need to also set `use_parent_material = true` on these nodes. 

![image](https://github.com/user-attachments/assets/106f4f72-78b9-4a62-9949-93fbd3079856)
_The nodes that were present at _ready have their property set, but the new runtime nodes do not_

But there is no great way to get references to these descendant nodes. 

- The signal `child_entered_tree` only works for direct children, not descendants. 
- When propagating, we can connect `child_entered_tree` on every child, and causing it to call another propagation on itself. This is a fine solution, but requires a fair bit of signal management. If these nodes ever get unparented from the Container, these signals need to be disconnected, etc. The moment I have to hook up hundreds of signals that are likely never to be used and all aim to do a single thing, I always feel like a better (more high level) solution must exist.
- A combination of `SceneTree`'s `node_entered_tree` signal and a function that checks `is_a_parent_of(node)` (or the Godot 4 equivalent `is_ancestor_of(node)`) is possible, but this would mean that for any node added anywhere in the tree, every single custom Container would perform this check. That does not seem desirable.

You could make the argument that since children get added to this container by the developer during runtime, it would be possible to manually set them to `use_parent_material = true` in the script where the `add_child()` call is. But not every script or node will know that they are a distant descendant of this custom Container. Additionally, there are nodes that get added on runtime by other built-in nodes (such as the `@VScrollbar` and `@HScrollbar` that get added to `ScrollContainers`.

So this leaves me with no convenient way to detect whether a node has been added to the tree, as a descendant of a node, without an somewhat volatile setup.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add an new `descendant_entered_tree` signal or a `NOTIFICATION_DESCENDANT_ENTERED_TREE` notification that gets fired whenever a node has been added to a tree. This would allow for easy detection whenever setup (such as the propagation of a call) needs to happen.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

I suspect such a signal or notification doesn't yet exist due to performance reasons. This makes sense, since every newly added node sending a notification up through the entire tree could be expensive, and adding nodes is extremely common.

However, certain signals have `virtual methods` that allow users to turn them on for nodes that specifically need them. For example the signal `global_transform_changed` in `Node` requires `set_notify_transform(true)` to be called. Perhaps a similar solution could prevent unnecessary calls here.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

I suspect it will be used a fair bit. Manually connecting every child and maintaining them that way is possible, but feels excessive and cumbersome for the seemingly simple use.

### Is there a reason why this should be core and not an add-on in the asset library?

`Node` is core.

**Comment by Mickeon** [comment]
> I suspect such a signal or notification doesn't yet exist due to performance reasons. This makes sense, since every newly added node sending a notification up through the entire tree could be expensive, and adding nodes is extremely common.

Correct.

> However, certain signals have `virtual methods` that allow users to turn them on for nodes that specifically need them. For example the signal `global_transform_changed` in `Node` requires `set_notify_transform(true)` to be called. Perhaps a similar solution could prevent unnecessary calls here.

You probably mean `NOTIFICATION_TRANSFORM_CHANGED` notification in both **CanvasItem** and **Node3D** (**Spatial** in 3.x), which is another concept.
This `notify_transform` "property" is actually used and set up internally by a lot of 3D nodes, and is exposed just so the user can do the same.
In comparison, **At least currently**, there's no justification to implement this internally, as these nodes propagate their own data in a way that's way more efficient than using a signal.


> If this enhancement will not be used often, can it be worked around with a few lines of script?


The SceneTree emits the [`node_added`](https://docs.godotengine.org/en/3.6/classes/class_scenetree.html#class-scenetree-signal-node-added) signal whenever a node is added to it, regardless of the node's hierarchy. You can then check if the new node is a nested child of the subject node with [is_a_parent_of](https://docs.godotengine.org/en/3.6/classes/class_node.html#class-node-method-is-a-parent-of) ([is_ancestor_of](https://docs.godotengine.org/en/latest/classes/class_node.html#class-node-method-is-ancestor-of) in Godot 4.x)

**Comment by Arecher** [comment]
> You probably mean `NOTIFICATION_TRANSFORM_CHANGED` notification in both **CanvasItem** and **Node3D** (**Spatial** in 3.x), which is another concept.

You're right, I meant to refer to the notification, but mistakenly wrote it down as a signal.

> In comparison, At least currently, there's no justification to implement this internally, as these nodes propagate their own data in a way that's way more efficient than using a signal.

Would it be more efficient to make it a notification instead? And does that mean that using signals isn't efficient?

> The SceneTree emits the [`node_added`](https://docs.godotengine.org/en/3.6/classes/class_scenetree.html#class-scenetree-signal-node-added) signal whenever a node is added to it, regardless of the node's hierarchy. You can then check if the new node is a nested child of the subject node with [is_a_parent_of](https://docs.godotengine.org/en/3.6/classes/class_node.html#class-node-method-is-a-parent-of) ([is_ancestor_of](https://docs.godotengine.org/en/latest/classes/class_node.html#class-node-method-is-ancestor-of) in Godot 4.x)

That was one of the solutions I thought of as well. But to me it seemed less performant to perform multiple checks for every single node added to the SceneTree (especially since `is_a_parent_of()` sounds like it's just a for loop that goes through the tree), than to hook up a few hundred/thousand signals, that fire only when a child is added to that specific descendant. Am I wrong in that assumption?

**Comment by Mickeon** [comment]
> Would it be more efficient to make it a notification instead? And does that mean that using signals isn't efficient?

I, personally, can't speak on whether it would be more efficient to use notifications. Signals are efficient and you shouldn't be afraid to use them if they solve a concrete problem. It's emitting a lot of signals at once, for extremely common operations, that can be concerning.

> That was one of the solutions I thought of as well. But to me it seemed less performant to perform multiple checks for every single node added to the SceneTree

It's not wrong, but for this case it may not even matter. For your use-case, there may be just a few of your custom containers in the **SceneTree**, which really wouldn't affect performance much. If you are worrying about performance without anything concrete just yet, remember that [premature optimizations are the root of all evil](https://www.geeksforgeeks.org/premature-optimization/).

**Comment by Arecher** [comment]
Personally I don't think choosing the most performant out of two directly available options, that are equal in complexity, development time and even lines of code, can be described as 'premature optimization'. That's just common sense to me.

The only thing it requires is to have a basic understanding of how costly certain calls/utilities in the engine are on average, which can be hard to know for certain, which is why I asked if my assumption was wrong. Learning the best/most performant practices and keeping those in mind seems extremely useful for any developer, and the concept of 'premature optimization' seems to often crop up and dismiss any such learning.

Having said all that, it doesn't really apply to your reply, since you did in fact answer and explained the context for your answer. But I've seen premature optimization used too much as a standard answer for any performance question rather than a footnote, and it's a huge pet peeve of mine.

> It's not wrong, but for this case it may not even matter. 

Thanks! I'll just keep my current setup as is in that case. I think I still would have expected Godot to offer a better built-in solution for cases like this _(such as the, in my opinion, the proposed signal/notification)_, but this will do.

**Comment by willnationsdev** [comment]
> A combination of SceneTree's node_entered_tree signal and a function that checks is_a_parent_of(node) (or the Godot 4 equivalent is_ancestor_of(node)) is possible, but this would mean that for any node added anywhere in the tree, every single custom Container would perform this check. That does not seem desirable.

This is how Godot Engine would internally need to implement this feature anyway, except it'd be even more generic to account for arbitrary root nodes and other use cases (so any opportunity for optimization is thrown out the window). In contrast, implementing your own solution tailored for your needs (while seemingly roundabout) is a much more effective means of resolving this issue.

I will say, however, that you can already minimize the "damage" of the callback if you merely check whether the added node is a `Control` node first before making the `is_ancestor_of(node)` call. That'll rule out most potentially wasted matches.
